lib/parsers/json_parser/validator.rb

```diff
@@ -0,0 +1,318 @@
+
+  def initialize(io)
+    @io = io
+    @current_node = nil # :object, :array, :string, :literal
+    @parent_nodes = []
+    @current_state = :awaiting_root_node
+    @escape_next = false
+    @current_literal_size = 0
+    @pos = 0
+
+    @all_parsers = {}
+
+    @execution_stats = {
+      array: 0,
+      object: 0,
+      literal: 0,
+      string: 0
+    }
+
+    setup_transitions
+  end
+
+  def validate
+    char_reader = FormatParser::UTF8Reader.new(@io)
+
+    while (c = char_reader.read_char)
+      @pos += 1
+      parse_char c
+
+      # Halt validation if the sampling limit is reached.
+      if @pos >= MAX_SAMPLE_SIZE
+        raise JSONParserError, "Invalid JSON file" if @current_state == :awaiting_root_node
+        return false
+      end
+    end
+
+    # Raising error in case the EOF is reached earlier than expected
+    raise JSONParserError, "Incomplete JSON file" if @current_state != :closed
+    true
+  rescue  FormatParser::UTF8Reader::UTF8CharReaderError
+    raise JSONParserError, "Invalid UTF-8 character"
+  end
+
+  def stats(node_type)
+    @execution_stats[node_type]
+  end
+
+  private
+
+  def setup_transitions
+    when_its :awaiting_root_node, ->(c) {
+      read_whitespace(c) or
+        start_object(c) or
+        start_array(c)
+    }
+
+    when_its :awaiting_object_attribute_key, ->(c) {
+      read_whitespace(c) or
+        start_attribute_key(c) or
+        close_object(c)
+    }
+
+    when_its :reading_object_attribute_key, ->(c) {
+      close_attribute_key(c) or
+        read_valid_string_char(c)
+    }
+
+    when_its :awaiting_object_colon_separator, ->(c) {
+      read_whitespace(c) or
+        read_colon(c)
+    }
+
+    when_its :awaiting_object_attribute_value, ->(c) {
+      read_whitespace(c) or
+        start_object(c) or
+        start_array(c) or
+        start_string(c) or
+        start_literal(c)
+    }
+
+    when_its :awaiting_array_value, ->(c) {
+      read_whitespace(c) or
+        start_object(c) or
+        start_array(c) or
+        start_string(c) or
+        start_literal(c) or
+        close_array(c)
+    }
+
+    when_its :reading_string, ->(c) {
+      close_string(c) or
+        read_valid_string_char(c)
+    }
+
+    when_its :awaiting_next_or_close, ->(c) {
+      read_whitespace(c) or
+        read_comma_separator(c) or
+        close_object(c) or
+        close_array(c)
+    }
+
+    when_its :reading_literal, ->(c) {
+      read_valid_literal_char(c) or (
+        close_literal(c) and (
+          read_whitespace(c) or
+          read_comma_separator(c) or
+          close_array(c) or
+          close_object(c)))
+    }
+
+    when_its :closed, ->(c) {
+      read_whitespace(c)
+    }
+  end
+
+  def when_its(state, act)
+    @all_parsers[state] = act
+  end
+
+  def parse_char(c)
+    next_step = @all_parsers[@current_state]
+    accepted = next_step.call(c)
+    reject_char(c) unless accepted
+  end
+
+  def read_whitespace(c)
```

Comment:
I would go with `?` methods here, given they return a bool. Looks more Ruby-like 😊

Reply:
@linkyndy 
Do you mean rename it to `read_whitespace?`  or replace this with the existing `whitespace?` method?

The reason having `?` seemed off to me was that these methods usually have side effects, which I don't expect a `?` method to have. (For instance,  `read_colon` updates @current_state.
These methods are more about performing an action and reporting success than actually answering a question.

I can see `read_whitespace?`, `read_valid_string_char` or `close_array` working, for example, but I wouldn't go for replacing with `whitespace?` altogether.

Wdyt?